File: app/policy_engine.py
"""
Policy Engine
-------------
Turns a risk score into a decision, given one agent's contract (its config
in config/agents.json). Two things are deterministic and NEVER overridden by
a risk score:
  1. blocked_action_types - a hard no for this agent, regardless of context.
  2. hard_constraints - physical/technical limits checked against the
     action's own parameters (e.g. rpm, temperature_c). This is the
     "Safety Engine" layer: the LLM can ask for anything, but a request
     that violates a wired-in limit is refused before it ever reaches
     the machine.

Everything else is risk-based: amount above the agent's autonomous ceiling,
or total risk score above its human-approval threshold, routes to a human
via an escalation instead of a hard block.
"""

from typing import Any, Dict, Optional, Tuple


def check_hard_constraints(hard_constraints: Dict[str, float], parameters: Optional[Dict[str, Any]]) -> list:
    violations = []
    parameters = parameters or {}
    for key, max_val in (hard_constraints or {}).items():
        if key in parameters:
            val = parameters[key]
            try:
                if float(val) > float(max_val):
                    violations.append(f"{key}={val} dépasse la limite câblée ({max_val})")
            except (TypeError, ValueError):
                continue
    return violations


def evaluate(
    agent_cfg: Dict[str, Any],
    action_type: str,
    amount: Optional[float],
    risk_result: Dict[str, Any],
    parameters: Optional[Dict[str, Any]],
) -> Tuple[str, str]:
    """Returns (decision, reason). decision is one of:
    'blocked' | 'pending_approval' | 'approved'
    """

    if action_type in agent_cfg.get("blocked_action_types", []):
        return "blocked", f"Type d'action '{action_type}' interdit pour cet agent (politique)."

    violations = check_hard_constraints(agent_cfg.get("hard_constraints", {}), parameters)
    if violations:
        return "blocked", "Contrainte physique/technique violée : " + "; ".join(violations)

    max_autonomous = agent_cfg.get("max_autonomous_amount")
    if amount is not None and max_autonomous is not None and amount > max_autonomous:
        return (
            "pending_approval",
            f"Montant {amount:g} € au-dessus du seuil autonome de cet agent ({max_autonomous:g} €).",
        )

    threshold = agent_cfg.get("required_human_approval_above", 100)
    if risk_result["risk_score"] > threshold:
        return (
            "pending_approval",
            f"Score de risque {risk_result['risk_score']} au-dessus du seuil de validation humaine ({threshold}).",
        )

    return "approved", "Action dans les limites du contrat de cet agent."

File: src/policy_engine.ts
import { AgentConfig, RiskResult, PolicyRule } from './types.js';

export function checkHardConstraints(hardConstraints: Record<string, number> = {}, parameters?: Record<string, any> | null): string[] {
  const violations: string[] = [];
  const params = parameters || {};
  for (const [key, maxVal] of Object.entries(hardConstraints || {})) {
    if (key in params) {
      const parsedVal = Number(params[key]); const parsedMax = Number(maxVal);
      if (!isNaN(parsedVal) && !isNaN(parsedMax) && parsedVal > parsedMax) violations.push(`${key}=${params[key]} exceeds safety limit (${maxVal})`);
    }
  }
  return violations;
}

function evaluateCondition(condition: PolicyRule['conditions'][0], context: { amount: number | null | undefined; action_type: string; risk_score: number; parameters: Record<string, any> }): boolean {
  let targetVal: any;
  if (condition.field === 'amount' || condition.field === 'refund_amount' || condition.field === 'wire_amount') targetVal = context.amount ?? (context.parameters[condition.field] != null ? Number(context.parameters[condition.field]) : undefined);
  else if (condition.field === 'action_type') targetVal = context.action_type;
  else if (condition.field === 'risk_score') targetVal = context.risk_score;
  else if (condition.field in context.parameters) targetVal = context.parameters[condition.field];
  if (targetVal === undefined || targetVal === null) return false;
  const expectedNum = Number(condition.value), targetNum = Number(targetVal), isNumeric = !isNaN(expectedNum) && !isNaN(targetNum);
  switch (condition.operator) {
    case '<=': return isNumeric && targetNum <= expectedNum;
    case '>=': return isNumeric && targetNum >= expectedNum;
    case '<': return isNumeric && targetNum < expectedNum;
    case '>': return isNumeric && targetNum > expectedNum;
    case '==': return isNumeric ? targetNum === expectedNum : String(targetVal).toLowerCase() === String(condition.value).toLowerCase();
    case '!=': return isNumeric ? targetNum !== expectedNum : String(targetVal).toLowerCase() !== String(condition.value).toLowerCase();
    case 'in': return String(condition.value).split(',').map(s => s.trim().toLowerCase()).includes(String(targetVal).toLowerCase());
    case 'contains': return String(targetVal).toLowerCase().includes(String(condition.value).toLowerCase());
    default: return false;
  }
}

/** Fail-closed autonomous control: actions are either approved or blocked. There is no human approval queue. */
export function evaluate(agentCfg: AgentConfig, actionType: string, amount: number | null | undefined, riskResult: RiskResult, parameters?: Record<string, any> | null, activePolicies: PolicyRule[] = [], agentId?: string): [decision: 'blocked' | 'pending_approval' | 'approved', reason: string, policyTriggered: string] {
  if (agentCfg.status && agentCfg.status !== 'active') return ['blocked', `Agent status is '${agentCfg.status.toUpperCase()}'. All actions are quarantined and blocked.`, 'Agent Lifecycle Security Gate'];
  const violations = checkHardConstraints(agentCfg.hard_constraints, parameters);
  if (violations.length) return ['blocked', `Hardware safety limit exceeded: ${violations.join('; ')}`, 'Physical Hard Limits'];
  if (agentCfg.blocked_action_types?.includes(actionType)) return ['blocked', `Action type '${actionType}' is forbidden for this agent.`, 'Agent Tier Boundary Policy'];

  const context = { amount, action_type: actionType, risk_score: riskResult.risk_score, parameters: parameters || {} };
  const matchingPolicies = activePolicies.filter(p => p.status === 'active' && (p.targetAgents.includes('*') || (!!agentId && p.targetAgents.includes(agentId))));

  for (const pol of matchingPolicies) for (const cond of pol.conditions) {
    if ((cond.action === 'BLOCK' || cond.action === 'HUMAN_REVIEW') && evaluateCondition(cond, context)) {
      return ['blocked', cond.action === 'HUMAN_REVIEW' ? `Action exceeds the policy boundary: ${cond.field} ${cond.operator} ${cond.value}. Human approval is not available.` : `Violated rule: ${cond.field} ${cond.operator} ${cond.value} under policy '${pol.name}'`, pol.name];
    }
  }

  const maxAutonomous = agentCfg.max_autonomous_amount;
  if (amount != null && maxAutonomous != null && amount > maxAutonomous) return ['blocked', `Amount €${amount.toLocaleString()} exceeds the agent autonomous limit of €${maxAutonomous.toLocaleString()}.`, 'Financial Threshold Safeguard'];

  const threshold = agentCfg.required_human_approval_above;
  if (threshold != null && riskResult.risk_score > threshold) return ['blocked', `Composite risk score ${riskResult.risk_score} exceeds the configured risk boundary of ${threshold}.`, 'Risk Boundary Policy'];

  for (const pol of matchingPolicies) for (const cond of pol.conditions) if (cond.action === 'ALLOW' && evaluateCondition(cond, context)) return ['approved', `Explicitly authorized by policy condition under '${pol.name}'.`, pol.name];
  return ['approved', 'Action verified within agent operational parameters.', 'Autonomous Policy Matrix'];
}

export function testPolicyRule(policy: PolicyRule, testPayload: { action_type: string; amount?: number | null; parameters?: Record<string, any> | null; risk_score?: number }): { matched: boolean; matched_condition?: PolicyRule['conditions'][0]; simulated_decision: 'ALLOW' | 'BLOCK' | 'HUMAN_REVIEW' | 'NO_MATCH'; reason: string } {
  const context = { amount: testPayload.amount ?? (testPayload.parameters?.amount ? Number(testPayload.parameters.amount) : undefined), action_type: testPayload.action_type, risk_score: testPayload.risk_score ?? 15, parameters: testPayload.parameters || {} };
  for (const cond of policy.conditions) if (evaluateCondition(cond, context)) return { matched: true, matched_condition: cond, simulated_decision: cond.action, reason: `Matched condition: ${cond.field} ${cond.operator} ${cond.value} => ${cond.action}` };
  return { matched: false, simulated_decision: 'NO_MATCH', reason: 'None of the candidate policy conditions matched the test payload.' };
}
